Code/Tools/FBuild/FBuildCore/Graph/ObjectNode.cpp

```diff
@@ -2021,6 +2021,10 @@ bool ObjectNode::WriteTmpFile( Job * job, AString & tmpDirectory, AString & tmpF
 
     FileStream tmpFile;
     AStackString<> fileName( sourceFile->GetName().FindLast( NATIVE_SLASH ) + 1 );
+    if (GetFlag( FLAG_GCC ))
+    {
+       fileName += ".i";
+    }
```

Comment:
Can you tell me a bit about what problems you had with this?  FASTBuild is built with gcc on Linux and I've not encountered this problem (though i have seen this in the past). It would be good to ensure we have test coverage for any case that requires this.

Reply:
The problem here is strictly connected to g++ being unable to distinguish between cpp file and preprocessed cpp file when file extension is .cpp. If it's the case then some defines automagically inserted by g++ at the beginning of the file (compiler defaults) are mistreated as regular macro definitions causing compilation to fail. The fix is to help compiler detect file content by adding proper extension (*.i). Seems like g++ 7 on my laptop is unaffected by this but on my friends machine it was a clear error (g++ 5.3.x if I remember correctly). I don't think this can be covered by unit testing though. The change is not harmful in any way and it works perfectly on g++ 7 and 5.3.x.

Reply:
I am using gcc 4.9.2 and am not seeing this problem. I wonder if it's how gcc is being invoked that is the problem - in all my cases, i'm compiling with: ```/<path>/gcc -std=c++11``` Can you tell me what args you are using, and if possible if this makes a difference on your version(s)?

If this is the issue, it would be nice to make this just work with a better preprocessed file extension along the lines of what you have proposed, but I think we need to detect .c vs .cpp and use .i or .ii respectively to be correct.